Use Compare trait like std::sort

lorenzhs opened the pull request:
You can now pass your own comparison operator to change sort order

this does seem to generate somewhat different code:

``` asm
  2.17 │260:   mov    %r8d,%r11d
  3.62 │       cmp    0x60(%rsp,%r11,4),%r15d
  0.89 │       setg   %r11b
  2.68 │       movzbl %r11b,%r11d
  0.93 │       lea    (%r11,%r8,2),%r8d
  1.98 │       mov    %edi,%r11d
  2.45 │       cmp    0x60(%rsp,%r11,4),%r14d
  0.67 │       setg   %r11b
  2.77 │       movzbl %r11b,%r11d
```

etc (old) vs new:

``` asm
  1.59 │2a0:   mov    %r9d,%ebx   
  1.28 │       mov    -0x14(%rax),%r11d
  2.93 │       cmp    %r14d,-0x450(%rbp,%rbx,4)
  1.84 │       setl   %bl
  1.84 │       movzbl %bl,%ebx 
  1.43 │       lea    (%rbx,%r9,2),%r9d    
  1.14 │       mov    %r8d,%ebx  
  2.61 │       cmp    %r11d,-0x450(%rbp,%rbx,4)      
  1.21 │       mov    -0x10(%rax),%r11d      
  1.52 │       setl   %bl                         
  1.81 │       movzbl %bl,%ebx          
  1.34 │       lea    (%rbx,%r8,2),%r8d  
```

maybe it's because I used `operator>(a, b)` before and now use `compare(b, a)` and that's confusing it?

// cc @Morwenn

lorenzhs committed:
Use Compare trait like std::sort

You can now pass your own comparison operator to change sort order

Morwenn commented:
Yeah, changing the operator might change the generated output. I can't say for sure though, I will try to have a look at it tonight.

lorenzhs commented on ssssort.h:
```diff
@@ -357,8 +363,12 @@ void ssssort_int(InputIterator begin, InputIterator end,
+          typename Compare = std::less<>,
+          typename std::enable_if<
+              std::is_convertible<bool, std::result_of_t<Compare(value_type, value_type)>>::value
+          >::type* = nullptr>
```
what a beast!

lorenzhs commented:
Nope it isn't the capture (just checked), but it's probably a good idea to capture that by value anyway.

Morwenn commented:
I tried to pass `compare` to every function instead of storing it in `Classifier`. The generated assembly is smaller and looks much more like the one without comparisons, but it's still different...

lorenzhs commented:
cool, can you send a PR for the compare branch?

Morwenn committed:
Don't store Compare in Classifier + more tidbits.

This commit ensures that a comparison function is always passed to every
function, but never explicitly stored. It seems to generate a better
assembly output.

There are several other minor things: less template parameters, more local
type aliases instead. I also removed the explicit passing of template
parameters to functions since they were not needed. That, and I turned the
std::enable_if_t into static assertions that should give more
understandable error messages.

lorenzhs commented:
It still generates the exact same assembly for me (g++ 5.4.0 on Debian unstable), huh.

Morwenn commented:
Eh, that's strange, it already shaved many instructions when I check on [Godbolt](https://gcc.godbolt.org/), even though g++ and libstdc++ didn't make it as perfect as it could have been.

lorenzhs commented:
Which compiler did you check with? Can you give me a link to the compiler explorer sample? (top left, near the sample selector)

Morwenn commented:
I checked g++ 6.1 and clang++ 3.8.

---

There may be a not-too-invasive solution to solve the libstdc++ problem: instead of calling `std::sort`, call the following function from your generic code:

    template<typename Iterator, typename Compare>
    void standard_sort(Iterator first, Iterator last, Compare compare) {
        std::sort(first, last, compare);
    }

    template<typename Iterator>
    void standard_sort(Iterator first, Iterator last, std::less<>) {
        std::sort(first, last);
    }

That way it should always call the `std::sort` variant that doesn't take a comparison function when given `std::less<>`, which is what we defaulted the algorithm to, and call the overload that uses the passed function object otherwise.

lorenzhs commented:
That doesn't make any difference for me, g++ is smart enough to optimize the iterator compare operator away. I would've been astonished if it wasn't, to be honest.

lorenzhs commented:
Here's a compiler explorer sketch to play around with the classifier: https://godbolt.org/g/MFPAcl

lorenzhs committed:
Add std::sort wrapper to prevent Compare object from being wrapped

Thanks, @Morwenn

Morwenn commented:
It seems that it did encode the source into the URL, but it's unable to test it for some reason :/
Could you instead create a Gist with the test code to make sure that we experiment on the same thing?

It's the right email, but you didn't need to add me :)

lorenzhs commented:
Sure: https://gist.github.com/lorenzhs/f795834ff1c1e7c47c634bbe11447728

Well you've done quite a bit, I think it's only fair to add you ;)

Morwenn commented:
Playing around with the classifier sketch, it seems that there is indeed a difference between `operator<` and `operator>`, but the version using `operator<` ends up generating 158 lines of assembly vs. 170 for the one using `operator>`. And the version using `compare` generates exactly the same code that the one using `operator<`, so... I guess that we aren't losing anything with `compare`.

Clang 3.8 seems to generate exactly the same code for the sketch, no matter whether we use `operator<`, `operator>` or `compare`. Looks good enough to me :)

Oh, and I'm helping partly because I plan to steal the code at some point and incorporate it into one of my libraries :p

lorenzhs commented on benchmark.h:
```diff
@@ -95,10 +95,11 @@ void run(T* data, const T* const copy, T* out, size_t size, Sorter sorter,
                  size_t outer_its, size_t inner_its,
-                 std::ofstream *stat_stream, bool deterministic_gen = false) {
+                 std::ofstream *stat_stream, bool deterministic_gen = false,
+                 Compare compare = Compare()) {
```
commenting so I remember to change this into `Compare compare = {}`

lorenzhs committed:
Restore C++11 compatibility

lorenzhs committed:
Ugly SFINAE hack to make g++ emit more efficient code

lorenzhs commented:
Thanks again, @Morwenn - I finally figured it out. It's really just gcc being stupid. See the `enable_if` SFINAE hackery on `step()`...

Morwenn commented:
Maybe it has problems determining whether some things will alias? I really don't know. I guess that a generic solution to this kind of problem would be to use something like `boost::call_traits<T>::param_type` which picks the best type for a read-only parameter (`const` reference or copy).

Anyway, glad I could help and it's cool that you managed to find and fix the problem :)

lorenzhs commented:
Well, the same worked fine when comparisons where done with `operator>` instead of `std::less<>` (with reversed arguments). I think it's a problem in the gcc optimizer, for it generated the more efficient code when I reversed the arguments to `compare` (with `setl` instead of `setg`, of course).

Not sure if/where to report that behaviour, though.

Morwenn commented:
I don't know about the *if*, but I sure know about *where*: you can report it on [GCC Bugzilla](https://gcc.gnu.org/bugzilla/) with the « missed optimization » tag (or something like that). I did that once, but the optimization was never implemented (though somebody confirmed the missed optimization a month ago or so).

lorenzhs commented:
Filed at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77614 - thanks again @Morwenn

Morwenn commented:
My pleasure :)
